HeterogeneousCore/CUDACore/interface/CUDAScopedContext.h

```diff
@@ -0,0 +1,252 @@
+#ifndef HeterogeneousCore_CUDACore_CUDAScopedContext_h
+#define HeterogeneousCore_CUDACore_CUDAScopedContext_h
+
+#include <optional>
+
+#include "CUDADataFormats/Common/interface/CUDAProduct.h"
+#include "FWCore/Concurrency/interface/WaitingTaskWithArenaHolder.h"
+#include "FWCore/Framework/interface/Event.h"
+#include "FWCore/Utilities/interface/EDGetToken.h"
+#include "FWCore/Utilities/interface/EDPutToken.h"
+#include "FWCore/Utilities/interface/StreamID.h"
+#include "HeterogeneousCore/CUDACore/interface/CUDAContextState.h"
+#include "HeterogeneousCore/CUDAUtilities/interface/SharedEventPtr.h"
+#include "HeterogeneousCore/CUDAUtilities/interface/SharedStreamPtr.h"
+
+namespace cudatest {
+  class TestCUDAScopedContext;
+}
+
+namespace impl {
+  // This class is intended to be derived by other CUDAScopedContext*, not for general use
+  class CUDAScopedContextBase {
+  public:
+    int device() const { return currentDevice_; }
+
+    // cudaStream_t is a pointer to a thread-safe object, for which a
+    // mutable access is needed even if the CUDAScopedContext itself
+    // would be const. Therefore it is ok to return a non-const
+    // pointer from a const method here.
+    cudaStream_t stream() const { return stream_.get(); }
+    const cudautils::SharedStreamPtr& streamPtr() const { return stream_; }
+
+  protected:
+    // The constructors set the current device device, but the device
```

Comment:
`current device device`, was that intended?

Reply:
I think not :-)

Reply:
One `device` too much indeed, will fix.